File: algo/29. Divide Two Integers.js
// Binary Search?

/**
 * @param {number} dividend
 * @param {number} divisor
 * @return {number}
 */
var divide = function(dividend, divisor) {
    var isNeg = dividend<0 != divisor<0; //? 0 : 1
    dividend = Math.abs(dividend);
    divisor = Math.abs(divisor);
    var left = dividend;
    var base = 1;
    var toSubtract = divisor;
    var ans = 0;
    while (left >= toSubtract) {
        left -= toSubtract;
        ans += base;
        base += base;
        toSubtract += toSubtract;
        
        // when the result is really close
        // we have to slow down the decreasing velocity
        if (left < toSubtract) {
            while (left >= divisor) {
                left -= divisor;
                ans += 1;
            }
        }
    }
    
    if (isNeg) {
        return Math.max(-ans, -Math.pow(2,31));
    } else {
        return Math.min(ans, Math.pow(2,31)-1);
    }
            
};

console.log(divide(-2147483648,1));

File: round2/29. Divide Two Integers.cs
public class Solution {
    public int Divide(int dividend, int divisor) {
        bool isNegative = (dividend < 0) != (divisor < 0);
        long absDividend = Math.Abs((long) dividend);
        long absDivisor = Math.Abs((long) divisor);
        long res = 0;
        long left = absDividend;
        long toSubstract = absDivisor;
        long toAdd = 1;
        while (left >= toSubstract) {
            left -= toSubstract;
            res += toAdd;
            toSubstract += toSubstract;
            toAdd += toAdd;
             
            if (left < toSubstract) {
                while (left >= absDivisor) {
                    res += 1;
                    left -= absDivisor;
                }
            }
        }
        if (isNegative) {
            return (int) Math.Max(-res, (long) int.MinValue);
        } else {
            return (int) Math.Min(res, (long) int.MaxValue);
        }
    }
}